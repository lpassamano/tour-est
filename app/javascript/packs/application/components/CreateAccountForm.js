import React, { Component } from "react";
import { navigate } from "@reach/router";

// TODO
// see if any other fields are necessary for the form!
// 2. add in note about password requirements (10 characters min)
// 3. add in field for password errors
// 4. add prop types

class CreateAccountForm extends Component {
  static defaultProps = { navigate };
  state = {
    username: "",
    password: "",
    password_confirmation: "",
    cultural_center: ""
  };

  handleSubmit = async event => {
    event.preventDefault();
    const result = await this.props.onCreateUser({
      username: this.state.username,
      password: this.state.password,
      cultural_center: this.state.cultural_center
    });
    console.log(result);
    if (result.ok) {
      this.props.navigate("/");
    } else {
      console.error(result.data.error);
    }
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          name="username"
          type="text"
          id="username"
          value={this.state.username}
          onChange={this.handleChange}
        />
        <input
          name="password"
          type="password"
          id="password"
          value={this.state.password}
          onChange={this.handleChange}
        />
        <input
          name="password_confirmation"
          type="password"
          id="password_confirmation"
          value={this.state.password_confirmation}
          onChange={this.handleChange}
        />
        <input
          name="cultural_center"
          type="text"
          id="cultural_center"
          value={this.state.cultural_center}
          onChange={this.handleChange}
        />
        <button type="submit">Log In</button>
      </form>
    );
  }
}

export default CreateAccountForm;
