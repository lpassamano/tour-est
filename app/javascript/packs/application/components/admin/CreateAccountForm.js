import React, { Component } from "react";
import { navigate } from "@reach/router";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import * as staffUserActions from "../../redux/staffUser/actions";

export class CreateAccountForm extends Component {
  state = {
    username: "",
    password: "",
    password_confirmation: "",
    cultural_center: ""
  };

  handleSubmit = event => {
    event.preventDefault();
    this.props.onCreateUser({
      user: {
        username: this.state.username,
        password: this.state.password,
        password_confirmation: this.state.password_confirmation
      },
      cultural_center: {
        name: this.state.cultural_center
      }
    });
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label htmlFor="username">Username: </label>
        <input
          name="username"
          type="text"
          value={this.state.username}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="password">Password: </label>
        <input
          name="password"
          type="password"
          value={this.state.password}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="password_confirmation">Confirm Password: </label>
        <input
          name="password_confirmation"
          type="password"
          value={this.state.password_confirmation}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="cultural_center">Organization: </label>
        <input
          name="cultural_center"
          type="text"
          value={this.state.cultural_center}
          onChange={this.handleChange}
        />
        <br />
        <button type="submit">Log In</button>
      </form>
    );
  }
}

CreateAccountForm.propTypes = {
  onCreateUser: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  onCreateUser: staffUserActions.createStaffUser
};

const enhance = connect(
  null,
  mapDispatchToProps
);

export default enhance(CreateAccountForm);
