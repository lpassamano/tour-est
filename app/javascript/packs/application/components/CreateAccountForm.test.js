import CreateAccountForm from "./CreateAccountForm";
import { shallow, mount } from "enzyme";
import React from "react";

describe("<CreateAccountForm />", () => {
  const setup = props => {
    const component = mount(<CreateAccountForm {...props} />);
    const event1 = { target: { name: "username", value: "leigh" } };
    const event2 = { target: { name: "password", value: "12345abcde" } };
    const event3 = { target: { name: "cultural_center", value: "Met Museum" } };
    component.find("input#username").simulate("change", event1);
    component.find("input#password").simulate("change", event2);
    component.find("input#cultural_center").simulate("change", event3);
    return component;
  };

  test("onSubmit - when submit is clicked the current values of username, password, and cultural center are submitted", done => {
    const onCreateUser = jest.fn();
    const navigate = to => {
      expect(to).toEqual("/");
      done();
    };
    onCreateUser.mockResolvedValue({ ok: true });
    const component = setup({ onCreateUser, navigate });
    component.find("form").simulate("submit", { preventDefault: () => null });
    expect(onCreateUser).toHaveBeenCalledWith({ username: "leigh", password: "12345abcde", cultural_center: "Met Museum"});
  });

  test("handleChange - when text is typed into the email, password, and cultural center fields the state is updated", () => {
    const component = setup({});
    expect(component.state("username")).toEqual("leigh");
    expect(component.state("password")).toEqual("12345abcde");
    expect(component.state("cultural_center")).toEqual("Met Museum");
  });

  test("render", () => {
    // creates a snapshot of the HTML that is rendered and if it changes the tests will fail
    // update the snapshot yarn test -u
    // google jest snapshot for more info
    const component = setup({});
    expect(component).toMatchSnapshot();
  });
});
