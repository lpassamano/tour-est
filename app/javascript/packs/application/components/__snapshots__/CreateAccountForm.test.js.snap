// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateAccountForm /> render 1`] = `
<CreateAccountForm
  onCreateUser={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="username"
    >
      Username: 
    </label>
    <input
      name="username"
      onChange={[Function]}
      type="text"
      value="leigh"
    />
    <br />
    <label
      htmlFor="password"
    >
      Password: 
    </label>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value="12345abcde"
    />
    <br />
    <label
      htmlFor="password_confirmation"
    >
      Confirm Password: 
    </label>
    <input
      name="password_confirmation"
      onChange={[Function]}
      type="password"
      value="12345abcde"
    />
    <br />
    <label
      htmlFor="cultural_center"
    >
      Organization: 
    </label>
    <input
      name="cultural_center"
      onChange={[Function]}
      type="text"
      value="Met Museum"
    />
    <br />
    <button
      type="submit"
    >
      Log In
    </button>
  </form>
</CreateAccountForm>
`;
